rfc/JENKINS-41334-parallel-stage-execution.md

```diff
@@ -0,0 +1,86 @@
+## RFC: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334) - Parallel stage execution
+
+* Author: Andrew Bayer
+* Created Date: February 27, 2017
+* Approved Date: n/a
+* JIRA: [JENKINS-41334](https://issues.jenkins-ci.org/browse/JENKINS-41334)
+* Status: Draft
+
+### Problem Statement
+
+The `parallel` step does not fit smoothly into Declarative. It doesn't
+allow for parallel stage execution, it doesn't utilize the `agent`
+definition, etc. There needs to be the ability to run `stage`s in
+parallel, with the full power of the Declarative syntax enabled -
+i.e., `agent`, `post`, `when`, etc.
+
+### Syntax
+
+#### Proposed example
+
+```groovy
+stage('foo') {
+    parallel {
+        stage('first') {
+            steps {
+                echo "First branch"
+            }
+        }
+        stage('second') {
+            steps {
+                echo "Second branch"
+            }
+        }
+    }
+}
+```
+
+#### Details
+
+* A `stage` has to contain one and only one of `steps` or
+    `parallel`. They cannot be combined in the same `stage`.
+* `stage` configuration can be done on both a `stage` containing
+    `parallel`, or `stage`s within `parallel`, i.e., `agent`, `when`,
+    `post`, etc.
+* `agent`, `environment`, and `tools` specified on the "parent"
+    `stage` will apply to child `stage`s in the same way that top-level
+    configuration applies to `stage`s currently.
+* Naming of `parallel` is up for debate.
+
+### Runtime Implementation
+
+There are two options for runtime. Which one we go with depends on
+Blue Ocean visualization decisions. The first option is to simply nest
+stage executions, while the second option is to not actually use the
+`stage` step for the nested `stage`s and instead utilize the `stage`
+names as the names of parallel branches.
+
+These two options each have their pros and cons. A pro in one option
+tends to lead to a con in the other option.
+
+#### Pro/Con
+
+* Stage status/metadata (i.e., skipped due to failure/`when`) would
```

Comment:
@abayer So, here's the scoop, after looking closer at the history and context:  using these tags is okay *as long as you are also generating appropriate ErrorActions, etc from running the pipeline.*  

I *think* unless we missed something critical at the time that the old way of using these was to *assist* on top of that mechanism by adding some additional context. 

What it *sounds* like is that there's a possibility of replacing the original flow structure and action information with the Metadata, for functional purposes.  And that is very incompatible. 

The flow needs to either execute correctly, or we need to make an upstream change in workflow-api to support use of status tags on nodes/blocks as a superset of just the current mechanisms.